### PR TITLE
reduce hand-IK coupling to hip offset

### PR DESCRIPTION
**Pull request by AndrewMeadows**
Fixed a bug where the inverse kinematics for the hands would move the hips too much when in HMD mode.

### CODE REVIEW

File: libraries/animation/src/AnimInverseKinematics.cpp [hunk 1]
```diff
@@ -144,168 +144,34 @@ void AnimInverseKinematics::solveWithCyclicCoordinateDescent(const std::vector<I
...
-                    lowestMovedIndex = pivotIndex;
-                }
-
-                // keep track of tip's new transform as we descend towards root
-                tipPosition = jointPosition + deltaRotation * leverArm;
-                tipOrientation = glm::normalize(deltaRotation * tipOrientation);
-                tipParentOrientation = glm::normalize(deltaRotation * tipParentOrientation);
-
-                pivotIndex = pivotsParentIndex;
-                pivotsParentIndex = _skeleton->getParentIndex(pivotIndex);
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: The big chunk of code above has moved into a new protected method called: **solveTargetWithCCD()**.

File: libraries/animation/src/AnimInverseKinematics.cpp [hunk 2]
```diff
@@ -641,6 +672,10 @@ void AnimInverseKinematics::initConstraints() {
             const float MAX_SPINE_SWING = PI / 14.0f;
             minDots.push_back(cosf(MAX_SPINE_SWING));
             stConstraint->setSwingLimits(minDots);
+            if (0 == baseName.compare("Spine1", Qt::CaseInsensitive)
+                    || 0 == baseName.compare("Spine", Qt::CaseInsensitive)) {
+                stConstraint->setLowerSpine(true);
+            }
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: This is where we flag the constraint as being part of the **lowerSpine**.

File: libraries/animation/src/AnimInverseKinematics.cpp [hunk 3]
```diff
@@ -329,6 +195,169 @@ void AnimInverseKinematics::solveWithCyclicCoordinateDescent(const std::vector<I
...
+                if (twistAxisLength > MIN_AXIS_LENGTH) {
+                    // project leverArm and targetLine to the plane
+                    twistAxis /= twistAxisLength;
+                    leverArm -= glm::dot(leverArm, twistAxis) * twistAxis;
+                    targetLine -= glm::dot(targetLine, twistAxis) * twistAxis;
+                } else {
+                    leverArm = Vectors::ZERO;
+                    targetLine = Vectors::ZERO;
+                }
+            }
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: Above is where we special-case the rotation of **lowerSpine** joints.  If we are solving **RotationAndPosition** or **HipsRelativeRotationAndPosition** targets (which is what the hands use (also the feet, but those don't pass through the spine joints)) then when we encounter **lowerSpine** joints we only allow the joints to twist -- no swing.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 14ec1b62954073b8d218da2af34ed9d768c22149 available at: https://gist.github.com/78698a0a3e6793c6c721<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6987.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6987.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: de54a0ac4bc86c867b693b168eb6702af2853d3a available at: https://gist.github.com/78698a0a3e6793c6c721<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6987.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6987.dmg

**Comment by AndrewMeadows** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: de54a0ac4bc86c867b693b168eb6702af2853d3a available at: https://gist.github.com/78698a0a3e6793c6c721<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6987.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6987.dmg

**Comment by PhilipRosedale** [comment]
Works great for me with and without hand controllers, good to go!

File: libraries/animation/src/AnimInverseKinematics.cpp (hunk 2 above)
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: For the record: @hyperlogic would like to see these string comparisons use **CaseSensitive**.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 71664dffae19d6703ce02035c38ed6b617b444d3 available at: https://gist.github.com/78698a0a3e6793c6c721<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6987.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6987.dmg

**Pull request merged**